gpu4pyscf/qmmm/pbc/itrf.py

```diff
@@ -0,0 +1,1131 @@
+                    Tijabc = cp.einsum('ijabc,ij->ijabc', Tijabc, invr7)
+                    Tijabc -= 8/5 / np.sqrt(np.pi) * cp.einsum('j,ij,ija,bc->ijabc', eta**5, ekR, R, np.eye(3))
+                    Tijabc -= 8/5 / np.sqrt(np.pi) * cp.einsum('j,ij,ijb,ac->ijabc', eta**5, ekR, R, np.eye(3))
+                    Tijabc -= 8/5 / np.sqrt(np.pi) * cp.einsum('j,ij,ijc,ab->ijabc', eta**5, ekR, R, np.eye(3))
+                    return Tija, Tijab, Tijabc
+
+            #------ qm - mm clasiical ewald energy gradient ------#
+            all_mm_coords = (mm_coords[None,:,:] - Lall[:,None,:]).reshape(-1,3)
+            all_mm_charges = cp.hstack([mm_charges] * len(Lall))
+            dist2 = all_mm_coords - cp.mean(qm_coords, axis=0)[None]
+            dist2 = cp.einsum('jx,jx->j', dist2, dist2)
+            if with_mm:
+                mm_ewovrl_grad = np.zeros_like(all_mm_coords)
+            mem_avail = cupy_helper.get_avail_mem()
+            blksize = int(mem_avail/64/3/len(all_mm_coords))
+            if blksize == 0:
+                raise RuntimeError(f"Not enough GPU memory, mem_avail = {mem_avail}, blkszie = {blksize}")
+            for i0, i1 in lib.prange(0, mol.natm, blksize):
+                R = qm_coords[i0:i1,None,:] - all_mm_coords[None,:,:]
+                r = cp.linalg.norm(R, axis=-1)
+                r[r<1e-16] = cp.inf
+
+                # substract the real-space Coulomb within rcut_hcore
+                mask = dist2 <= cell.rcut_hcore**2
+                Tija, Tijab, Tijabc = grad_Tij(R[:,mask], r[:,mask])
+                qm_ewovrl_grad[i0:i1] -= cp.einsum('i,ijx,j->ix', qm_charges[i0:i1], Tija, all_mm_charges[mask])
+                qm_ewovrl_grad[i0:i1] -= cp.einsum('ia,ijxa,j->ix', qm_dipoles[i0:i1], Tijab, all_mm_charges[mask])
+                qm_ewovrl_grad[i0:i1] -= cp.einsum('iab,ijxab,j->ix', qm_quads[i0:i1], Tijabc, all_mm_charges[mask]) / 3
+                if with_mm:
+                    mm_ewovrl_grad[mask] += cp.einsum('i,ijx,j->jx', qm_charges[i0:i1], Tija, all_mm_charges[mask])
+                    mm_ewovrl_grad[mask] += cp.einsum('ia,ijxa,j->jx', qm_dipoles[i0:i1], Tijab, all_mm_charges[mask])
+                    mm_ewovrl_grad[mask] += cp.einsum('iab,ijxab,j->jx', qm_quads[i0:i1], Tijabc, all_mm_charges[mask]) / 3
+
+                # difference between MM gaussain charges and MM point charges
+                mask = dist2 > cell.rcut_hcore**2
+                zetas = cp.asarray(cell.get_zetas())
+                min_expnt = cp.min(zetas)
+                max_ewrcut = pbc.gto.cell._estimate_rcut(min_expnt, 0, 1., cell.precision)
+                cut2 = (max_ewrcut + rmax_qm)**2
+                mask = mask & (dist2 <= cut2)
+                expnts = cp.hstack([cp.sqrt(zetas)] * len(Lall))[mask]
+                Tija, Tijab, Tijabc = grad_kTij(R[:,mask], r[:,mask], expnts)
+                qm_ewovrl_grad[i0:i1] -= cp.einsum('i,ijx,j->ix', qm_charges[i0:i1], Tija, all_mm_charges[mask])
```

Comment:
The same tensor contraction issue here. `cp.einsum` behaves differently from `np.einsum`.